### ISSUE TITLE
Notification Object Manual

### ISSUE THREAD

**Issue by Courtney3141** [seL4/seL4#414]
In your seL4 manual on page 25 it states: 

"If the notification word is zero at the time `seL4_Wait()` is called, the invoker blocks."

I have spoken to Curtis and apparently this is wrong. The blocking state does not care about the value of the badge.

**Comment by gernotheiser** [comment]
This describes the intended behaviour: the waiter is unblocked as soon as there's something to collect. If the word is non-zero to start with, the call should return immediately. If it's different then that would seem to be a bug. Unless I'm missing something. @xurtis?

**Comment by Courtney3141** [comment]
If the notification object has been signaled with a cap without a badge, Curtis told me that the notification state machine changes from the idle state to the active state, regardless of what the badge is. This (supposed) mistake revealed itself as I thought it was unusual that I first had to badge a notification object before I could expect behavior similar to a lock (the manual implies that the signal is lost in the case that the cap is unbadged, where as if the cap is badged the signal is stored in the form of the notification word)

**Comment by Courtney3141** [comment]
![notification_object_state_machine](https://user-images.githubusercontent.com/55342552/123511915-ad581900-d6c7-11eb-8c20-5866b7d66638.jpg)
This is the state machine of the notification object which Curtis described to me (with Waiting replaced with Blocking)

**Comment by lsf37** [comment]
> If the notification object has been signaled with a cap without a badge, Curtis told me that the notification state machine changes from the idle state to the active state, regardless of what the badge is. This (supposed) mistake revealed itself as I thought it was unusual that I first had to badge a notification object before I could expect behavior similar to a lock (the manual implies that the signal is lost in the case that the cap is unbadged, where as if the cap is badged the signal is stored in the form of the notification word)

That is correct, yes. The spec is here: https://github.com/seL4/l4v/blob/ee3b84fb57acb5653e44c3df1fbee000d9666529/spec/abstract/Ipc_A.thy#L464

The badge is data only, it is irrelevant for the blocking behaviour. The intended use case is to use badged notification caps for sending, because the badge will enable you to identify where the signal came from. If there is only one possible sender, there is nothing wrong with using an unbadged notification cap, though. For badged notification caps you get exactly the behaviour Gernot described.

**Comment by gernotheiser** [comment]
Yeah, I forgot that you can signal with an unbadged cap. As Gerwin says, this doesn't make sense in most circumstances, except where you use the Notification as a pure binary semaphore.

**Comment by kevinelp** [comment]
Just a comment, I'm not sure we're benefiting here by changing the simple description and model in the manual to a description of a state machine involving hidden implementation state. Making badge zero a NOP as originally intended preserves the simple description, does not introduce implementation details, and does not restrict use cases as far as I see. Signalling with badge zero seems like an edge case that could be avoided completely.   Binary semaphores are simply any non-zero badge. 

This is just me preferring converging on simple models, rather than exposing more complex ones.

**Comment by kevinelp** [comment]
Another thought, zero badge notification behaviour is "defined from a proof perspective, but unsupported, and should not be relied upon".  I wanted to use the typical "undefined" term meaning open to future definition, but that is probably more confusing given the verified software context.

**Comment by lsf37** [comment]
> Just a comment, I'm not sure we're benefiting here by changing the simple description and model in the manual to a description of a state machine involving hidden implementation state. Making badge zero a NOP as originally intended preserves the simple description, does not introduce implementation details, and does not restrict use cases as far as I see. Signalling with badge zero seems like an edge case that could be avoided completely. Binary semaphores are simply any non-zero badge.
>
> This is just me preferring converging on simple models, rather than exposing more complex ones.

I can see the preference for a simpler model, but if we do that, then what we're documenting is emergent behaviour, not the actual behaviour (of either spec of code). I agree that the emergent behaviour is as described in this case, but we have no proof for that. We do have proof that it behaves like the spec, which has always been a state machine with queues, and badges as data. All of these are relevant for the actual behaviour of the kernel. If you simplify them away, you will either not explain some behaviour (the wait/send queues) or you risk getting it slightly wrong, esp when the implementation and spec evolve.

We could explain both in the manual: a description of the actual mechanism underneath, which may or may not be used as intended, and then the intended use case which is simpler to understand.

**Comment by kevinelp** [comment]
I'd argue we have emergent specification and proof :-) I vaguely recall the introduction of the state machine for ease of proof under the condition that behaviour was the same, performance no worse . This looks more like an edge case that slipped in, got specified based on the implemented behaviour and proved. Not handling this case may have even been a performance optimisation at the time of the common case - don't waste cycles on erroneous edge cases, just let it fail later.

I'm curious, if unbadged notify was a NOP, wouldn't the simpler model specify the behaviour? 

Independent of the issue under discussion, a question arises of how do we handle rough edges in the API like this (if we agree this is a rough edge). We can document them in the hope people don't cut themselves, and/or we can smooth them over.  I'd like a strategy that does not encourage reliance on edge case behaviour and thus permits smoothing in the future should we ever get the resources to do house keeping. 

This might boil down to a use manual versus a reference manual, don't know. If there was an easy answer we'd have done it by now.

**Comment by xurtis** [comment]
> Another thought, zero badge notification behaviour is "defined from a proof perspective, but unsupported, and should not be relied upon".

This is a behavior that is assumed by all of our locking and semaphore primitives.

I'm also not personally a fan of describing the behavior of the notification objects in terms of the value in the data word, my experience has been that it is a less useful as model for using it to build systems around notifications that the state machine that is actually used to implement it.

**Comment by lsf37** [comment]
> I'd argue we have emergent specification and proof :-) I vaguely recall the introduction of the state machine for ease of proof under the condition that behaviour was the same, performance no worse . 

No, definitely not :-) The description of badges as locks is several years (at least 4-5) older than the state machine in both spec and code. It's entirely possible there was a discussion between Phil and you on that on the conceptual level before it made it into the Haskell prototype and was always planned as locks, but the proof side has never seen anything else, and it was not changed for performance. Remember these started out as async endpoints and we were thinking about them differently back then (maybe we shouldn't have, but that's different).

With the current implementation, the state machine and queues are required in the spec. The view of badges as locks does not describe the behaviour in enough detail to do things like infoflow (would have too much nondeterminism which could hide flows). If you do badge 0 as NOP, which would be entirely possible, we could maybe put another spec+proof on top which does have that nondeterminism. This would make that view more robust.

> I'm curious, if unbadged notify was a NOP, wouldn't the simpler model specify the behaviour?

I'm not actually certain that the description of badges as locks is simpler. It looks simpler in text, but mostly because people "know" what a lock is and therefore you don't have to describe that much. If you actually fully describe everything, it looks to me like at least the same level of complexity. In the end, a lock is also a state machine, but in the badge view, you'd have to describe multiple state machines (one for each bit), and you then haven't yet dealt with badges that have multiple bits set. Whereas now, there is one state machine which `or`s badges as pure data.

> Independent of the issue under discussion, a question arises of how do we handle rough edges in the API like this (if we agree this is a rough edge). We can document them in the hope people don't cut themselves, and/or we can smooth them over. I'd like a strategy that does not encourage reliance on edge case behaviour and thus permits smoothing in the future should we ever get the resources to do house keeping.

In general I'd agree that we want to make things smoother over time, but I don't fully agree that this is a rough edge case. It mostly depends on perspective, i.e. what you see as the underlying mechanism. I've never seen this as anything else than a state machine with data, and am perfectly content with that. I don't find it difficult. I don't find the view/use as lock difficult either, but for me it's one (the main) use case of a slightly more general mechanism.

There is probably a lot of code out there that does use this feature, and I'd argue that it is both useful and clear for the case of exactly one sender and receiver, which is reasonably common. You don't have to deal with badges, deriving caps, or even ever looking at the badge word. You just wait and get woken up. If that's the only thing you need, it's perfectly simple. 
 
> This might boil down to a use manual versus a reference manual, don't know. If there was an easy answer we'd have done it by now.

Yeah, true :-)

**Comment by kevinelp** [comment]
Fandom or not is probably partly a function of familiarity with the implementation and local libraries.

I agree that for sync primitives,  the value is largely irrelevant. Any non-zero value  suffices in the original model, and zero works as well with the current implementation. 

What I take from this is that the behaviour is already relied up by some.

**Comment by lsf37** [comment]
> Fandom or not is probably partly a function of familiarity with the implementation and local libraries.
> 
> I agree that for sync primitives, the value is largely irrelevant. Any non-zero value suffices in the original model, and zero works as well with the current implementation.

Agreed, with both. I'm not trying to say the implementation mechanism is better, I just don't think it's worse.
 
> What I take from this is that the behaviour is already relied up by some.

Yes. It's been about 10 years ;-)

**Comment by kevinelp** [comment]
The original model was not really "locks". It was a semaphore-like construct with wait blocking when the value was zero or resetting the value, and signal was an 'or' operator. It could be used as a binary semaphore or select or whatever..

I'm curious as to what the description of the state machine might be in the manual as I find the current state labels counterintuitive. Notification objects are passive, "Idle", "Waiting", "Active" immediately invoke connotations of threads and my head hurts when trying to reconcile the two.

**Comment by kent-mcleod** [comment]
As of today, the manual still describes a model inconsistent with the implementation: https://github.com/seL4/seL4/blob/master/manual/parts/notifications.tex#L38. So at least that should be brought in line.
